sbc_compassion/models/correspondence.py

```diff
@@ -389,26 +391,98 @@ def write(self, vals):
+                # for translation. Just keep the original page.
+                final_pdf.addPage(existing_pdf.getPage(i))
+                continue
+            # Take the boxes depending on which page we handle
+            if i == 0:
+                boxes = layout.page_1_box_ids
+            elif i == 1:
+                boxes = layout.page_2_box_ids
+            else:
+                boxes = layout.additional_page_box_id
+            if not boxes:
+                remaining_text = text
+                continue
+            box_texts = text.split(BOX_SEPARATOR)
+            if len(box_texts) > len(boxes):
+                # There should never be more text than expected by the
+                # layout.
+                return False
+
+            # Construct new PDF for the current page
+            page_output = PdfFileWriter()
+            page_output.addPage(existing_pdf.getPage(i))
+            if i > 1:
+                # Add the original page to the final pdf. Translation will
+                # come in separate page.
+                final_pdf.addPage(page_output.getPage(0))
+            # Compose the text for each box inside the page
+            for j in xrange(0, len(box_texts)):
+                text = remaining_text + box_texts[j]
+                remaining_text = ''
+                box = boxes[j]
+                # Check that the text can fit in the box
+                if len(text) > box.text_length:
+                    # Log when text is too long to see if that happens a lot
+                    self.message_post(
+                        'Translation went out of the translation box',
+                        'Translation too long')
+                    trunc = text[:box.text_length]
+                    remaining_text = text[box.text_length:]
+                    text = trunc
+
+                translation_pdf = box.get_pdf(text)
+                if i <= 1:
+                    # Merge the translation on the existing page
+                    page = page_output.getPage(j)
+                    page.mergePage(translation_pdf.getPage(0))
+                    page_output.addPage(page)
+                else:
+                    # Add translation in a separate page
+                    final_pdf.addPage(translation_pdf.getPage(0))
+
+            # Write the last version of the page into final pdf
+            if i <= 1:
+                final_pdf.addPage(page_output.getPage(j))
+
+        # Add page if there is remaining text
+        if remaining_text:
+            translation_pdf = layout.additional_page_box_id.get_pdf(
+                remaining_text)
+            final_pdf.addPage(translation_pdf.getPage(0))
+
+        # Finally write the pdf back into letter_image
+        output_stream = BytesIO()
+        final_pdf.write(output_stream)
+        output_stream.seek(0)
+        self.letter_image.datas = base64.b64encode(output_stream.read())
+
         return True
```

Comment:
I hope the code is understable, if not don't hesitate to make some comments or questions that can help improve.